### PR TITLE
Add AMD GPU support for cpp codegen

### PR DESCRIPTION
**Pull request by jataylo**
Adds support for hsaco binaries in inductor's cpp_wrapper codegen and enables the CUDA tests in test_cpp_wrapper.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by jataylo** [comment]
cc: @dllehr-amd

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -141,7 +141,14 @@ def codegen(self, code: IndentedBuffer, device_cm_stack: contextlib.ExitStack):
...
             if self.first_time:
-                code.writeline(f"at::cuda::CUDAGuard device_guard({self.device_idx});")
+                if torch.version.hip is None:
+                    code.writeline(
+                        f"at::cuda::CUDAGuard device_guard({self.device_idx});"
+                    )
+                else:
+                    code.writeline(
+                        f"at::hip::HIPGuard device_guard({self.device_idx});"
+                    )
```
Comment (jithunnair-amd) [inline torch/_inductor/codegen/wrapper.py:151]: @jataylo I'll need to dig further into the history to figure out why, but hipify actually has the following mapping for `CUDAGuard`:
`("CUDAGuard", ("HIPGuardMasqueradingAsCUDA", API_PYTORCH)),`
Did you try that already and run into some issues?

File: torch/_inductor/codegen/wrapper.py
Comment (jithunnair-amd) [inline torch/_inductor/codegen/wrapper.py:1]: @jataylo Actually, *most* of the changes in this file can be handled by hipify correctly, so I'd recommend trying the following:
1. Add the following line to `tools/amd_build/build_amd.py`:
```
     ignores=ignores,
+    extra_files=["torch/_inductor/codegen/wrapper.py"],
     out_of_place_only=args.out_of_place_only,
```

2. Only retain changes from this changeset that aren't handled by hipify eg.:
```
<< AT_ERROR("CUDA driver error: ", static_cast<int>(__err));
---
>> AT_ERROR("HIP driver error: ", static_cast<int>(__err));
```
and the cubin-path-related changes.

**Review by jithunnair-amd — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/wrapper.py (hunk 1 above)
Comment (jataylo) [reply torch/_inductor/codegen/wrapper.py]: Initially I had manually tried to use HIPGuardMasqueradingAsCUDA but ran into some issues, likely a mistake on my part. With the hipify integration this works as expected.

**Comment by jataylo** [comment]
Hi @malfet @kit1980, opening this up for review now, could you help take a look?

**Comment by jataylo** [comment]
This commit https://github.com/pytorch/pytorch/commit/44fdfd3222243e766370218cb4a4d1c77435249c made a lot of changes and this PR will have to be refreshed. Moving to draft now.

cc: @dllehr-amd @jithunnair-amd

**Comment by jataylo** [comment]
Creating a new PR https://github.com/pytorch/pytorch/pull/105141 as the upstream implementation has undergone a few changes making this stale.

**Pull request closed without merging**